Ruby scripting support

remi6397 opened the pull request:
Replaces #1006 

This PR adds support for the [Ruby](https://www.ruby-lang.org) programming language, or to be precise a lightweight implementation of Ruby called [MRuby](https://mruby.org/).

## Resources

- [MRuby API documentation](https://mruby.org/docs/api/)

___

<details>
<summary>pre-0.90 Description</summary>
This PR adds basic support for [Ruby](https://www.ruby-lang.org) scripting via [MRuby](https://mruby.org/).

I thought it'd be nice to have Ruby as an option for scripting, because it's widely used in WebDev ([Rails](https://rubyonrails.org/), [Sinatra](http://sinatrarb.com/), etc.), DevOps ([Puppet](https://puppet.com), [Chef](https://www.chef.io), [Vagrant](https://www.vagrantup.com/)), ITSec ([Logstash](https://www.elastic.co/logstash), [Metasploit](https://www.metasploit.com/)) and probably in many more applications beyond my field of expertise (SketchUp being the first one to come to my mind).

If you decided to merge, Ruby would make the second most popular language of those available in TIC (right after JavaScript).

I've already tested this on Windows (10, x64) and Linux (openSUSE Tumbleweed) and it compiles without any problems. There are still ~macOS~, Emscripten and Android left to test, but: ~1) I'd have to install XCode my Mac;~ 2) I don't know how to x-compile for Emscripten; 3) I don't own any Android device.

Apart from that, everything seems to work flawlessly. :wink:

EDIT: Ran build on my Mac, it works there too.
</details>

joshgoebel commented on CMakeLists.txt:
```diff
@@ -300,12 +373,13 @@ macro(MACRO_CORE SCRIPT DEFINE BUILD_DEPRECATED)
         ${TIC80CORE_DIR}/api/fennel.c 
         ${TIC80CORE_DIR}/api/wren.c 
         ${TIC80CORE_DIR}/api/squirrel.c
+        ${TIC80CORE_DIR}/api/mrb.c
```
Could we name this `mruby` since that seems to be the name used everywhere else?

joshgoebel commented on README.md:
```diff
@@ -18,6 +18,7 @@ To make a retro styled game, the whole process of creation and execution takes p
 - Multiple programming languages: [Lua](https://www.lua.org),
   [Moonscript](https://moonscript.org),
   [Javascript](https://developer.mozilla.org/en-US/docs/Web/JavaScript),
+  [Ruby](https://www.ruby-lang.org/en/),
```
I think this should be https://mruby.org

MRuby and Ruby are NOT quite the same project.  It's certainly possible to embed Ruby also, but since we're not doing that we should be clear that we're using Ruby, at the very _least_ the link here should point to MRuby I'd think.

remi6397 replied:
Yes, I've been thinking of that too, there is some source-level incompatibility between the two, but at the same time we shouldn't create any more confusion to the end users.

IMO it's something worth mentioning in the documentation.

joshgoebel replied:
> we shouldn't create any more confusion to the end users.

I think `[Ruby](https://mruby.org)` would solve the issue, no?

remi6397 replied:
> I think [Ruby](https://mruby.org) would solve the issue, no?

Ruby is the language, MRuby is just one of the available implementations.

Just like with Javascript (or should I say ECMAscript); duktape isn't perfectly compliant either:

> ```markdown
>    [Javascript](https://developer.mozilla.org/en-US/docs/Web/JavaScript), 
> ```

joshgoebel commented on src/studio/demos.c:
```diff
@@ -88,6 +88,29 @@ tic_script_config_extra JsSyntaxConfigExtra =
+    #include "../build/assets/rubymark.tic.dat"
+};
+
+tic_script_config_extra RubySyntaxConfigExtra =
```
Should this be `MRubySyntaxConfigExtra`?

joshgoebel commented:
This is MUCH cleaner PR. :)  Nice!

remi6397 commented:
> This is MUCH cleaner PR. :) Nice!

Yet at the same time people who were watching the previous discussion are now MIA :upside_down_face:

joshgoebel commented:
You mentioned it in the thread, they can follow if they desire.

joshgoebel commented:
You could always have just force pushed this to the old PR branch (with or without any rebasing)...

remi6397 commented:
> You could always have just force pushed this to the old PR branch (with or without any rebasing)...

That's what I've done before, and if you look at that branch on my profile, it doesn't show all those duplicate commits. Seems it's just the PR system that tends to break. I might be wrong though.

RobLoach commented:
This is looking much cleaner, nicely done! My only concern on this change is that `yac` and `rake` become build dependencies for TIC-80. You can no longer build TIC-80 without them installed on your system. They may not be available on smaller platforms like Dingoo or the 68000, and it will break those builds.

Would love to see if we could get in to work without the added dependencies 🤔

bggd commented:
mruby can be built without ```yacc```(or GNU Bison) after mruby 2.1.2+( https://mruby.org/releases/2020/08/06/mruby-2.1.2-released.html).

nesbox commented:
I tried to delete `find_program(YACC yacc)` line and TIC80 was built without errors, seems YACC detection can be deleted.

nesbox commented:
> They may not be available on smaller platforms like Dingoo or the 68000, and it will break those builds.

Do these platforms support cross-compilation, as we do with RPi baremetal build for example?

remi6397 commented:
@nesbox I think that'd be it, wouldn't it? https://github.com/nesbox/TIC-80/pull/1726/commits/0ea4ecaf832b5e80ef6aef6bef234ba4f85fdac9

nesbox commented:
> Apropos cross compilation, will you be able to fix your cross docker images soon?

I tried but without success, I removed your `ruby/setup-ruby` action and added installation commands directly to the container
```
    - name: Install Ruby
      run: |
        apt-get install ruby-full -y
        ruby --version
```
and got these errors
https://github.com/nesbox/TIC-80/runs/4468004547?check_suite_focus=true#step:5:876
```
sh: 1: gcc: not found
rake aborted!
Command failed with status (127): [gcc -E -P -std=gnu99 -g -O3 -Wall -Wundef ...]
/__w/TIC-80/TIC-80/vendor/mruby/lib/mruby/build/command.rb:37:in `_run'
/__w/TIC-80/TIC-80/vendor/mruby/lib/mruby/build/command.rb:99:in `run'
/__w/TIC-80/TIC-80/vendor/mruby/lib/mruby/build/command.rb:120:in `block (2 levels) in define_rules'
```
As I understand, RAKE doesn't see `gcc` from the cross-compile location.

remi6397 commented:
> > Apropos cross compilation, will you be able to fix your cross docker images soon?
> 
> I tried but without success, I removed your `ruby/setup-ruby` action and added installation commands directly to the container
> 
> ```
>     - name: Install Ruby
>       run: |
>         apt-get install ruby-full -y
>         ruby --version
> ```
> 
> and got these errors https://github.com/nesbox/TIC-80/runs/4468004547?check_suite_focus=true#step:5:876
> 
> ```
> sh: 1: gcc: not found
> rake aborted!
> Command failed with status (127): [gcc -E -P -std=gnu99 -g -O3 -Wall -Wundef ...]
> /__w/TIC-80/TIC-80/vendor/mruby/lib/mruby/build/command.rb:37:in `_run'
> /__w/TIC-80/TIC-80/vendor/mruby/lib/mruby/build/command.rb:99:in `run'
> /__w/TIC-80/TIC-80/vendor/mruby/lib/mruby/build/command.rb:120:in `block (2 levels) in define_rules'
> ```
> 
> As I understand, RAKE doesn't see `gcc` from the cross-compile location.

 @nesbox Looks like that error occured in mruby/build/host. Is the **\*host\*** toolchain present?

bggd commented:
> ```Command failed with status (127): [gcc -E -P -std=gnu99 -g -O3 -Wall -Wundef ...]```

It seems to use ```gcc``` instead of ```arm-linux-gnueabihf-gcc```(/opt/tools/arm-bcm2708/arm-linux-gnueabihf/bin/arm-linux-gnueabihf-gcc) on rpi.

nesbox commented:
finally, I successfully compiled ruby support for RPi

<img width="152" alt="image" src="https://user-images.githubusercontent.com/1101448/145559337-42396168-e524-4f84-b934-4b44cb9d629a.png">

the host toolchain changes:
``` yml
    - name: Install Host toolchain
      run: |
        apt-get update
        apt-get install --assume-yes build-essential ruby-full
```
and add `-std=gnu99` to the `CMakeLists.txt`
``` cmake
if(RPI)
    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -std=gnu99")
endif()
```

nesbox commented:
Also, I doubt it's even possible to build MRUBY for RpiBaremetal and N3DS because there is no `sys/socket.h` at all.
<img width="666" alt="image" src="https://user-images.githubusercontent.com/1101448/145560456-1ccd80b8-5b8a-4cbc-9647-3ef0b2aeb470.png">.
I think we could merge PR without MRuby on those platforms, only major platforms should be supported (`win linux mac rpi android`), so only Android remains to be fixed for now.

remi6397 commented:
> Also, I doubt it's even possible to build MRUBY for RpiBaremetal and N3DS because there is no `sys/socket.h` at all.

Well, we can always skip building that gem for baremetalpi ;)

nesbox commented:
@remi6397 
Another question, can we be sure that the network and file I/O are disabled in MRUBY for the TIC-80?

remi6397 commented:
> @remi6397 Another question, can we be sure that the network and file I/O are disabled in MRUBY for the TIC-80?

Of course we can skip building all undesired **mrbgems**, such as those that interface with the system. Just keep in mind that you can never really be safe without an actual sandbox. Check out chromium sources for ideas.

nesbox commented:
> Of course we can skip building all undesired mrbgems, such as those that interface with the system. Just keep in mind that you can never really be safe without an actual sandbox. Check out chromium sources for ideas.

Thanks, I'm worried not about security, but about TIC-80 limitations, because carts shouldn't have access to the file system and the network :)

remi6397 commented:
> Thanks, I'm worried not about security

Phew, that's comforting _/j_

I personally jail programs that execute third-party code, but sometime in the future it might be worth to consider native sandboxing.

nesbox commented:
@remi6397 I see N3DS is fixed, how did you do it?
<img width="137" alt="image" src="https://user-images.githubusercontent.com/1101448/145606652-63f945ce-a099-4b38-8cb0-e2a9fb214bc8.png">

remi6397 commented:
> @remi6397 I see N3DS is fixed, how did you do it? <img alt="image" width="137" src="https://user-images.githubusercontent.com/1101448/145606652-63f945ce-a099-4b38-8cb0-e2a9fb214bc8.png">

That's because I got rid of IO-related gems, I think.

nesbox commented:
Please try to add `$(TIC80LIB)/libmruby.a \`
here https://github.com/remi6397/TIC-80/blob/master/build/baremetalpi/Makefile#L23

```makefile
LIBS :=  \
	$(TIC80LIB)/libtic80studio.a \
	$(TIC80LIB)/libtic80core.a \
	$(TIC80LIB)/libgiflib.a \
	$(TIC80LIB)/liblpeg.a \
 	$(TIC80LIB)/liblua.a \
	$(TIC80LIB)/libzip.a \
	$(TIC80LIB)/libblipbuf.a \
	$(TIC80LIB)/libsquirrel.a \
	$(TIC80LIB)/libwren.a \
	$(TIC80LIB)/libargparse.a \
	$(TIC80LIB)/libwave_writer.a \
	$(TIC80LIB)/libduktape.a \
	$(TIC80LIB)/libzlib.a \
	$(TIC80LIB)/libpng.a \
	$(NEWLIBDIR)/lib/libc.a \
        $(NEWLIBDIR)/lib/libcirclenewlib.a \
	$(CIRCLEHOME)/addon/vc4/sound/libvchiqsound.a \
        $(CIRCLEHOME)/addon/vc4/vchiq/libvchiq.a \
        $(CIRCLEHOME)/addon/linux/liblinuxemu.a \
 	$(CIRCLEHOME)/addon/SDCard/libsdcard.a \
 	$(CIRCLEHOME)/addon/fatfs/libfatfs.a \
  	$(CIRCLEHOME)/lib/usb/libusb.a \
 	$(CIRCLEHOME)/lib/input/libinput.a \
 	$(CIRCLEHOME)/lib/fs/fat/libfatfs.a \
 	$(CIRCLEHOME)/lib/fs/libfs.a \
  	$(CIRCLEHOME)/lib/net/libnet.a \
  	$(CIRCLEHOME)/lib/sched/libsched.a \
  	$(CIRCLEHOME)/lib/libcircle.a \
	$(NEWLIBDIR)/lib/libm.a
```

This will possibly fix the rpi-baremetal build.

nesbox commented:
Left to win Android
<img width="619" alt="image" src="https://user-images.githubusercontent.com/1101448/145612032-1bd91d43-8c99-479a-ab1b-b90ecb00218f.png">

RobLoach commented:
Would love to keep backwards compatibility when Rake isn't available. Here's a Pull Request that disables mruby if Rake isn't there: https://github.com/remi6397/TIC-80/pull/6

remi6397 commented:
And Bob's your uncle, everything builds successfully now! :tada: 

![Build Status](https://user-images.githubusercontent.com/10615862/145680206-c548c21b-73a4-4f4f-9c45-34882705ffa6.png)

nesbox commented:
Awesome, so we're ready to merge :)
Thanks for your persistence and for not abandoning the PR 👍

nesbox commented on .github/workflows/build.yml:
```diff
@@ -454,6 +518,28 @@ jobs:
+        for ARCH in arm arm64 x86 x86_64; do \
+          API=24; \
+          SYSROOT=$ANDROID_NDK_HOME/platforms/android-$API/arch-$ARCH; \
+          python3 "$ANDROID_NDK_HOME/build/tools/make_standalone_toolchain.py" \
```
I'm going to move this part to the Docker initialization part after merging the PR

nesbox commented on .github/workflows/build.yml:
```diff
@@ -588,7 +678,7 @@ jobs:
     strategy:
       matrix:
-        script: [lua, moon, fennel, js, wren, squirrel]
+        script: [lua, moon, fennel, mruby, js, wren, squirrel]
```
We have to rename it to `ruby` here or the `export ruby %name%` command won't work.

nesbox replied:
Or, we can use `export mruby` as a variant.
Any thoughts?

nesbox commented on src/api/mruby.c:
```diff
@@ -0,0 +1,1233 @@
+            mrb_funcall(mrb, mrb_top_self(mrb), TicFunc, 0);
+            catcherr(machine);
+
+            if (mrb_respond_to(mrb, mrb_top_self(mrb), mrb_intern_cstr(mrb, OvrFunc)))
```
We don't need to call `OVR()` here for new languages, it's for compatibility, so we can remove this callback here.

nesbox reviewed (approved):
Thank you, the job well done 👍
